### PR TITLE
indexer: add catchup

### PR DESCRIPTION
**Pull request by hosted-fornet**
Add logic to allow the bootstrapping of new indexer from existing ones. Requires the new indexer be already connected to the %rollup app so that the batches can be verified as correct. Also made the "inflated state" real -- before there was no inflating of state.

Testing goes something like this:

```
::  on ~zod
.smart-lib/noun +zig!mk-smart
|install our %zig
:rollup|activate
:sequencer|init our 0x0 0xf3b2.f5ab.92df.a29d.5e8e.fe70.98cb.ed99.3856.949b.fb4f.5cfb.7bca.4a45.0ae7.0e50
:indexer &set-sequencer [our %sequencer]
:indexer &set-rollup [our %rollup]
:uqbar|set-sources our 0x0 our
:wallet &zig-wallet-poke [%populate 0xbeef]
:wallet &zig-wallet-poke [%submit-custom from=0x7772.b8a7.6840.8922.2903.5b28.7494.436f.8850.713c to=0x74.6361.7274.6e6f.632d.7367.697a town=0x0 gas=[1 1.000.000] args='[%give 0xc7ec.a38c.5c74.d58d.04b0.6650.4772.f3a6.d02e.92f5 `0x532c.d5cf.befc.5c0f.e5dd.3910.8f23.79fd 10.000.000.000.000.000.000 1.000]' my-grains=(silt ~[0x532c.d5cf.befc.5c0f.975c.8e60.c3f6.fcac]) cont-grains=(silt ~[0x532c.d5cf.befc.5c0f.e5dd.3910.8f23.79fd])]
:sequencer &sequencer-town-action [%trigger-batch ~]

::  on ~nec
.smart-lib/noun +zig!mk-smart
|install our %zig

::  on ~nec after ~zod has finished set up
:indexer &set-sequencer [~zod %sequencer]
:indexer &set-rollup [~zod %rollup]

::  on ~nec, to catchup indexer
:indexer &indexer-catchup [~zod %indexer]
```

Tested with `mule` in place of `zebra` in `mill.hoon`.

### CODE REVIEW

File: app/indexer.hoon [hunk 1]
```diff
@@ -1061,19 +726,43 @@
...
       ==
     ~
   [%hash combined-batch combined-egg combined-grain]
 ::
+++  inflate-state
+  |=  batches-by-town-list=(list [@ux =batches:ui batch-order:ui])
+  ^-  indices-0
+  =|  temporary-state=_state
+  |^
+  |-
```
Comment (tacryt-socryp) [inline app/indexer.hoon:741]: just a note: it doesn't change the semantics here, but `|^` introduces a `$` arm as well as `|-` here, so you can actually get rid of the `|-` without changing any of the other code and still be able to recurse.

**Review by tacryt-socryp — APPROVED** [review]
LGTM. Do indexers ever validate the data they're consuming, or do they just trust it? I don't see any of that, though I could have missed it.

**Comment by hosted-fornet** [comment]
> LGTM. Do indexers ever validate the data they're consuming, or do they just trust it? I don't see any of that, though I could have missed it.

The model is that %indexer waits for confirmation that a batch root has been added by %rollup before indexing a new batch.

When we get a batch from a %sequencer, we:
1. if we have a corresponding batch root from %rollup, immediately index batch
2. else, stash batch until we get a confirmation from %rollup that the batch root has been added

See, e.g. [indexer lines 524-556](https://github.com/uqbar-dao/ziggurat/blob/6a25e63cc3f81b966f060cc1c4de9840c6d5cc09/app/indexer.hoon#L524-L556). The handling of updates from %rollup 

We do a similar set of steps when we get a batch root from %rollup:
1. if we have a corresponding batch from %sequencer, immediately index that batch
2. else, stash the batch root until we get a corresponding batch from a %sequencer

See, e.g. [indexer lines 485-516](https://github.com/uqbar-dao/ziggurat/blob/6a25e63cc3f81b966f060cc1c4de9840c6d5cc09/app/indexer.hoon#L485-L516)

File: app/indexer.hoon (hunk 1 above)
Comment (hosted-fornet) [reply app/indexer.hoon]: Ah, good catch. Removed unnecessary `|-` in 9158285

**Pull request merged**